core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodeService.java

```diff
@@ -62,19 +63,21 @@ public DiscoveryNodeService addCustomAttributeProvider(CustomAttributesProvider
     }
 
     public DiscoveryNode buildLocalNode(TransportAddress publishAddress) {
-        Map<String, String> attributes = new HashMap<>(settings.getByPrefix("node.").getAsMap());
+        final String nodeId = generateNodeId(settings);
+        return new DiscoveryNode(settings.get("node.name"), nodeId, publishAddress, buildAttributes(), version);
+    }
+
+    Map<String, String> buildAttributes() {
+        Map<String, String> attributes = new HashMap<>(Node.NODE_ATTRIBUTES.get(this.settings).getAsMap());
         attributes.remove("name"); // name is extracted in other places
         if (attributes.containsKey("client")) {
-            if (attributes.get("client").equals("false")) {
-                attributes.remove("client"); // this is the default
-            } else {
-                // if we are client node, don't store data ...
-                attributes.put("data", "false");
-            }
+            throw new IllegalArgumentException("node.client setting is no longer supported, use " + Node.NODE_MASTER_SETTING.getKey()
+                + ", " + Node.NODE_DATA_SETTING.getKey() + " and " + Node.NODE_INGEST_SETTING.getKey() + " explicitly instead");
         }
-        if (attributes.containsKey("data")) {
-            if (attributes.get("data").equals("true")) {
-                attributes.remove("data");
+        //nocommit why don't we remove master as well if it's true? and ingest?
+        if (attributes.containsKey(DiscoveryNode.Role.DATA.getRoleName())) {
```

Comment:
dude this entire method makes no sense to me. Can we just make sure we make these attributes dense rather than sparse and just fill all the roles for every node?